Switch flash_map to resolve devices at compile-time instead of using device_get_binding

de-nordic opened the pull request:
The PR contains series of commits that switch flash_map from using device_get_binding on flash_area.fa_dev_name
to using new filed flash_area.fa_dev, that is filled at compile time with a pointer to the flash device the flash_area is tied to.
This PR also contains commit that makes the filed flash_area.fa_dev_name optional, as it is no longer used within flash map API, but is left available for external code that relies on existence of this field.

Set to DNM due to ongoing tests.

nvlsianpu commented on subsys/storage/flash_map/Kconfig:
```diff
@@ -21,6 +21,19 @@ config FLASH_MAP_SHELL
 	  This enables shell commands to list and test flash maps.
 
+
+config FLASH_MAP_DEV_NAME_IN_FLASH_AREA
```
Maybe for backward compatibility this option should impose device binding behavior in the code?
``` C
#ifdef CONFIG_FLASH_MAP_DEV_NAME_IN_FLASH_AREA
	if (fa->fa_dev == NULL) {
		flash_dev = device_get_binding(fa->fa_dev_name);
		if (flash_dev == NULL) {
			return -ENODEV;	
		}
	}

#endif
```

de-nordic replied:
I have such variant of the PR if desired. I have thought about the options and I now think, since that we allow custom flash_map to be developed, that user may fail to deliver device bindings so the option to use the new compile time resolution should be default for non custom maps.

nvlsianpu replied:
I agree with you.

nvlsianpu replied:
#ifdef CONFIG_FLASH_MAP_DEV_NAME_IN_FLASH_AREA ||
	if (fa->fa_dev == NULL) {
		flash_dev = device_get_binding(fa->fa_dev_name);
		if (flash_dev == NULL) {
			return -ENODEV;	
		}
	}
#endif
```

de-nordic committed:
storage/flash_map: Make flash_area.fa_dev_name optional

With switch to compile time resolved device object and usage of
flash_area.fa_dev, the filed flash_area.fa_dev_name becomes optional
as there are no more device_get_binding calls within flash_map
that rely on that filed, or any other references to that field.

The added Kconfig option, FLASH_MAP_DEV_NAME_IN_FLASH_AREA, allows
to remove the flash_area.fa_dev_name field, when disabled.
The option is set to y by default to support external code that may
relay on the existence of flash_area.fa_dev_name.

Configurations for samples and tests that require the fa_dev_name
have been altered to have proper Kconfig options enabled.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nvlsianpu reviewed (commented):
FLASH_MAP_DEV_IN_FLASH_AREA=y
FLASH_MAP_USE_DEV_IN_FLASH_AREA=n

makes `fa_dev` field meaningless. This doesn't make sense.

de-nordic commented:
> 
> 
> FLASH_MAP_DEV_IN_FLASH_AREA=y
> FLASH_MAP_USE_DEV_IN_FLASH_AREA=n
> 
> makes `fa_dev` field meaningless. This doesn't make sense.

That is in case when user wants to use fa_dev for something other than flash map API.

nvlsianpu commented:
> That is in case when user wants to use fa_dev for something other than flash map API.

Is there any user case? I understand that user might want either `fa_dev` and `fa_dev_name`. Worth to add possibility to use `fa_dev_name` if `fa_dev` is null (backward compatibility, and dynamic drivers).

I feel that so simple subsystem have to much option with this patch.

de-nordic commented:
> 
> Is there any user case?

I don't know. I will default the one choice: if `fa_dev` is enable it is used.

> I understand that user might want either `fa_dev` and `fa_dev_name`. Worth to add possibility to use `fa_dev_name` if `fa_dev` is null (backward compatibility, and dynamic drivers).

I am not sure I want to add such option: the selection should go either way, if use the `fa_dev`, then use `fa_dev` if not, then use `fa_dev_name`. We will end up in situations where user gets `fa_dev` null due to error in generating map and `fa_dev_name` points to something else than expected. We can only control the default flash map. No harm unless some writes happen.

> 
> I feel that so simple subsystem have to much option with this patch.

Will reduce the _USE_ one.

de-nordic committed:
storage/flash_map: Switch to using flash_area.fa_dev

The commit adds Kconfig option FLASH_MAP_DEV_IN_FLASH_AREA that
enables addition of const struct device * type fa_dev field to
the flash_area structure.
Enabling this option will also populate fa_dev of each entry of
the default flash map with a pointer to a device object that flash
area described by the entry is tied to.
The option switches flash map API to use the flash_area.fa_dev
instead of obtaining device object by name, flash_area.fa_dev_name,
by default.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nvlsianpu commented on subsys/fs/Kconfig.littlefs:
```diff
@@ -8,6 +8,7 @@ config FILE_SYSTEM_LITTLEFS
 	depends on FILE_SYSTEM
 	depends on FLASH_MAP
 	depends on FLASH_PAGE_LAYOUT
+	depends on FLASH_MAP_DEV_NAME_IN_FLASH_AREA
```
Only reason of forcing device_name is the line subsys/fs/littlefs_fs.c 638:
``` C
 LOG_ERR("can't get flash device: %s", log_strdup(fs->area->fa_dev_name));
```

I would replace it by:
``` C
LOG_ERR("can't get flash device of flash area id: %d", area_id);
```
and resign from this dependency.

nvlsianpu replied:
+ modify a log line in `samples/subsys/fs/littlefs/src/main.c`

de-nordic replied:
I left it there because this can be decided afterwards and wanted the minimal change in code outside of the flash API.

de-nordic committed:
fs: littlefs: Remove references to fa_dev_name

The flash subsystem API is switching from using fa_dev_name to fa_dev,
of flash_area structure, making the fa_dev_name optional filed,
controlled by Kconfig option CONFIG_FLASH_MAP_DEV_NAME_IN_FLASH_AREA.
Using fa_dev_name for logging makes it required field, so this commit
removed the usage from log to remove dependency on the Kconfig option.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
fs: samples: littlefs: Remove references to fa_dev_name

The flash subsystem API is switching from using fa_dev_name to fa_dev,
of flash_area structure, making the fa_dev_name optional filed,
controlled by Kconfig option CONFIG_FLASH_MAP_DEV_NAME_IN_FLASH_AREA.
Using fa_dev_name for logging makes it required field, so this commit
removed the usage from log to remove dependency on the Kconfig option.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nvlsianpu committed:
fs/fcb: Remove references to fa_dev_name

The flash subsystem API is switching from using fa_dev_name to fa_dev,
of flash_area structure, making the fa_dev_name optional filed,
controlled by Kconfig option CONFIG_FLASH_MAP_DEV_NAME_IN_FLASH_AREA.
Using fa_dev_name for device binding makes it required field,
so this commit introduce binding using flash_area_get_device() which
removes dependency on the Kconfig option.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests/subsys/fs/fcb: Remove references to fa_dev_name

The flash subsystem API is switching from using fa_dev_name to fa_dev,
of flash_area structure, making the fa_dev_name optional filed,
controlled by Kconfig option CONFIG_FLASH_MAP_DEV_NAME_IN_FLASH_AREA.
Using fa_dev_name for device binding makes it required field,
so this commit introduce binding using flash_area_get_device() which
removes dependency on the Kconfig option.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented on subsys/settings/Kconfig:
```diff
@@ -60,6 +60,7 @@ config SETTINGS_NVS
 	bool "NVS non-volatile storage support"
 	depends on NVS
 	depends on FLASH_MAP
+	depends on FLASH_MAP_DEV_NAME_IN_FLASH_AREA
```
Are the plane is to use the device reference in nvs API?  `int nvs_init(struct nvs_fs *fs, const char *dev_name) -> int nvs_init(struct nvs_fs *fs, const struct device *dev)`

nvlsianpu replied:
^^ not in this commit of course.

de-nordic replied:
Yeah. Actually I have been planning to leave the original API call and add alternative nvs_init taking the device as a param; then I would replace every instance of the nvs_init call within zephyr (modules, etc) with the new API call but leave the original one available, maybe deprecate it.
It might be called directly by user and I want to leave people a chance to adjust before removing it completely.

nvlsianpu replied:
maybe we can move this dependency to NVS itself.

de-nordic replied:
Yes, I will do that.

de-nordic replied:
Retreating. The nvs does not need the `fa_dev_name`, the callers need it: for example nvs basic sample directly provided name of device.

galak reviewed (commented):
Curious if you've thought of always having .fa_dev and changing the name logic to use the device name `.fa_dev->name`?

de-nordic commented:
> 
> 
> Curious if you've thought of always having .fa_dev and changing the name logic to use the device name `.fa_dev->name`?

Ha! Good point.

galak commented:
> > 
> 
> Ha! Good point. Haven thought about that.
> Although we leave the fa_dev_name for the custom maps that fail to generate fa_dev..

Figure we can save a bit on the ifdef'ery :).

de-nordic commented:
> 
> Figure we can save a bit on the ifdef'ery :).

We will probably dump the `fa_dev_name` completely but we have to deprecate it so that users that generate customs flash maps get used to generating pointers to device objects instead of device names.

Laczen commented:
@de-nordic, @nvlsianpu, can't we get rid of flash_map? Since all information can be received from the dts, I would expect this to be replaced by a `flash_partition_write`, `flash_partition_read` and `flash_partition_erase` that would additionally do a check to avoid doing a flash operation outside the area. A flash partition cfg could be defined that contains the offset and size and points to the flash cfg for erase_value, write_size and erase_block_size.

de-nordic commented:
> 
> 
> @de-nordic, @nvlsianpu, can't we get rid of flash_map? Since all information can be received from the dts, I would expect this to be replaced by a `flash_partition_write`, `flash_partition_read` and `flash_partition_erase` that would additionally do a check to avoid doing a flash operation outside the area. A flash partition cfg could be defined that contains the offset and size and points to the flash cfg for erase_value, write_size and erase_block_size.

Cannot yet. The msuboot and mcumgr use integer IDs for figuring which partition to operate on. Within Zephyr, internally we probably could do that.

Laczen commented:
> Cannot yet. The msuboot and mcumgr use integer IDs for figuring which partition to operate on. Within Zephyr, internally we probably could do that.

If required by mcuboot/mcumgr, these things should move into mcuboot/mcumgr code.

mbolivar-nordic commented on subsys/storage/flash_map/Kconfig:
```diff
@@ -18,8 +18,47 @@ if FLASH_MAP
+	  flash_area.fa_dev_name to get device by name.
+	  In case of the default flash map, enabling this option will cause the fa_dev, of
+	  each map entry, to be populated with a pointer to the device tied to that flash area.
+	  The option is selected for default flash map (FLASH_MAP_CUSTOM=n); if your custom map
+	  populates the flash_area.fa_dev, with pointers to devices that are tied to proper flash
+	  areas, you may want to unselect this option.
```
I don't understand this.

> The option is selected

Nothing I can see `select`s this option if FLASH_MAP_CUSTOM=n, it's just `default y` in that case. That is not the same thing.

> if your custom map [...] you may want to unselect this option.

I don't understand what this is trying to say, either.

First, I think using the word "unselect" in this context is unclear because Kconfig has a `select` keyword that you cannot undo.

Second, it seems to me that the only way to both have a custom map (`CONFIG_FLASH_MAP_CUSTOM=y`) and populate `fa_dev` pointers is if the user manually sets them up, so why would they want to make sure `CONFIG_FLASH_MAP_DEV_IN_FLASH_AREA=n`?

de-nordic replied:
I will fix the comments.

de-nordic replied:
Seems that I have merged comments from two different options.

mbolivar-nordic reviewed (changes requested):
-1 for now because I'm not comfortable with this at first blush and want to take a longer look at this today

mbolivar-nordic commented:
@de-nordic @nvlsianpu I'm confused about why this is being done this way. I feel like we've been carrying a bunch of legacy stuff from the mynewt operating system in Zephyr's mcuboot port that I don't understand the need for and this seems to be adding on top of it instead of fixing the underling issue.

From what I recall, `struct flash_area` is used in mcuboot basically as a copy/paste from the mynewt RTOS HAL. And now when I look at what's inside the mcuboot repository today, I see that each flash backend is responsible for defining the structure according to its own needs.

Grepping a bit more, I only see `fa_device_id` mentioned in 2 places in bootutil itself:

- https://github.com/mcu-tools/mcuboot/blob/579b30c29999860f9f7d843a25ff5453b6542cce/boot/bootutil/src/loader.c#L1947
- https://github.com/mcu-tools/mcuboot/blob/579b30c29999860f9f7d843a25ff5453b6542cce/boot/bootutil/src/loader.c#L2469

Those are write-only, just putting the value into the `struct boot_rsp`. And back in https://github.com/mcu-tools/mcuboot/blob/579b30c29999860f9f7d843a25ff5453b6542cce/boot/zephyr/main.c, it doesn't look like we're doing anything with that value at all.

So it seems to me we could do this in a couple of different ways that would have lower footprint.

**Option 1** Stick with the mynewt HAL names, but change `fa_device_id` to a `uintptr_t` and put the device pointer in there. We'd need to rearrange a bit to minimize padding bytes, so the results would look like this:

```c
struct flash_area {
	uintptr_t fa_device_id;
	off_t fa_off;
	size_t fa_size;
	uint8_t fa_id;
};
```

Then set `.fa_device_id = (uintptr_t)DEVICE_DT_GET(...)` in the C macro magic that initializes this at build time.

Zephyr is already using differently sized types for the fa_size and fa_off members (we use `size_t` and `off_t`, but MCUboot's PORTING.md says they are `uint32_t`, which isn't the same: `size_t` is bigger on 64 bit systems, and `off_t` is signed). So there's already some skew here in the Zephyr code, and we should be able to use a differently sized integer type for fa_device_id if we want to IMO.

**Option 2** Take this to upstream MCUboot and change it so this bootloader doesn't force design decisions made by the mynewt RTOS onto other operating systems.

From what I can tell, that would basically mean changing the `br_flash_dev_id` in `struct boot_rsp` to be an opaque typedef instead,

```c
struct boot_rsp {
    const struct image_header *br_hdr;
    flash_dev_id_t br_flash_dev_id;
    uint32_t br_image_off;
};
```

which Zephyr's flash map backend could handle like this:

```c
typedef const struct device* flash_dev_id_t;
```

Then our flash_area structure can use a device pointer here too, with nicer results since there will be less casting to and from `const struct device*`:

```
struct flash_area {
	flash_dev_id_t fa_device_id;
	off_t fa_off;
	size_t fa_size;
	uint8_t fa_id;
};
```

Am I missing something here? Why do we need to bloat this structure even more?

mbolivar-nordic commented:
> Am I missing something here? Why do we need to bloat this structure even more?

To head off a potential objection, yes, I know that what I am proposing is an API change and that [Flash map is stable](https://docs.zephyrproject.org/latest/reference/overview.html).

But I think we all agree that the current situation is just technical debt we inherited from another RTOS. We have a process for changing stable APIs and I think a change like this in particular is long overdue, especially now that we can avoid using device_get_binding() at all.

Laczen commented:
Maybe flash_map should be replaced by a flash_partition API that would limit the flash read, write, erase to a specific partition. This could even integrate the flash_page API by defining a erase_block_size at partition level. This could be done straight from the dts.

Flash_map as is could be moved to mcuboot.

nvlsianpu commented:
@mbolivar This PR objective is to use compile time device binding instead of runtime binding. Legacy fa_dev_name field is kept optional only for backward compatibility.

And yes, we should improve this code as you elaborated (I found both options usefully) and change the API.

We want to limit this PR to already small change. API change we would do in very next PR.

mbolivar-nordic commented:
> We want to limit this PR to already small change. API change we would do in very next PR.

I guess I don't understand why this PR should be merged if it's adding a new Kconfig option that will be removed in the next PR. If the path forward to using compile-time struct devices is agreed upon -- it seems like you agree, would be good to know @de-nordic's thoughts too -- then why not pursue it?

de-nordic commented:
The `flash_area` structure definition is not shared with mcuboot as far as I can tell. The only thing that is being exchanged between mcuboot/mcumgr and Zephyr is the ID (`fa_id`), and as a `flas_area*` API call parameters.
The `flash_area` structure is used internally only.
We could probably get rid of the `fa_device_id`, you may post your comment (https://github.com/zephyrproject-rtos/zephyr/pull/33346#issuecomment-809554075) as an issue.
I do not think that changing type of `fa_device_id` to let it take `const struct device *` casted to `uintptr_t` is good idea.

The ifdefery and Kconfig options are provided because we let users to link custom flash maps, which means that we have let them adapt what they generate and actually give them option whether they want to adapt at all (at least for now).
The objective of the PR has been to get rid of `device_get_binding` usage within Zephyr, internally yet allow alternative.


> 
> I guess I don't understand why this PR should be merged if it's adding a new Kconfig option that will be removed in the next PR.

I do not think that the Kconfig will be removed in next PR. The `fa_device_id` might be, Addition of data placeholders, like this PR does are quite safe, removals not always.
Yes we consider changes to `flash_map` API but we do not have consensus yet whether this will be change to existing API or alternative API that will phase out the original within Zephyr so that we could later on deprecate the original one.

mbolivar-nordic commented:
> The mcuboot should not directly peek into flash_area, I have thought that we are giving it flash_map header and it blindly uses `flash_area_* `API calls.

No, that's not correct. It needs the structure definition provided by a platform port. This is covered in https://github.com/mcu-tools/mcuboot/blob/master/docs/PORTING.md. You can see in flash_map.h for Zephyr that `fa_device_id` is provided directly for MCUboot compatibility; we don't use it anywhere in the tree except to print it out in the shell.

It's not just documentation either. The mcuboot code looks directly into the `flash_area` structure in several places in bootutil. By grep, it directly looks at `fa_id`, `fa_size`, `fa_device_id`, `fa_off` -- all of the members it mentions in `struct flash_area` in PORTING.md.

> We could probably get rid of the fa_device_id

I wish!

But since mcuboot is getting this definition from here, we can't do that without breaking mcuboot for the reason explained above.

Try it and see; if you comment it out in Zephyr and build mcuboot, you will get this:

```
/home/mbolivar/zp/bootloader/mcuboot/boot/bootutil/src/loader.c:1947:58: error: 'const struct flash_area' has no member named 'fa_device_id'
 1947 |     rsp->br_flash_dev_id = BOOT_IMG_AREA(state, BOOT_PRIMARY_SLOT)->fa_device_id;
```

de-nordic commented:
> 
> No, that's not correct. It needs the structure definition provided by a platform port. This is covered in https://github.com/mcu-tools/mcuboot/blob/master/docs/PORTING.md. You can see in flash_map.h for Zephyr that `fa_device_id` is provided directly for MCUboot compatibility; we don't use it anywhere in the tree except to print it out in the shell.
> 
Yes, it also states that you may need to implement counters to protect flash area from closing it when it has been opened more than one time. 
So it is bad design. The object of flash area should be only accessed an manipulated via API or macros that hide the internals of the structure. if you want something from it it should be taken by the API and the flash area structure should be implemented. 

> 
> > We could probably get rid of the fa_device_id
> 
> I wish!
> 
> But since mcuboot is getting this definition from here, we can't do that without breaking mcuboot for the reason explained above.
Then we should change mcuboot code to access the structure via API/macros only and make it take our flash area definition when compiled with Zephry

> 
> > The mcuboot should not directly peek into flash_area, I have thought that we are giving it flash_map header and it blindly uses `flash_area_* `API calls.
> 
> No, that's not correct. It needs the structure definition provided by a platform port. This is covered in https://github.com/mcu-tools/mcuboot/blob/master/docs/PORTING.md. You can see in flash_map.h for Zephyr that `fa_device_id` is provided directly for MCUboot compatibility; we don't use it anywhere in the tree except to print it out in the shell.

That was my wishful thinking.

> 
> It's not just documentation either. The mcuboot code looks directly into the `flash_area` structure in several places in bootutil. By grep, it directly looks at `fa_id`, `fa_size`, `fa_device_id`, `fa_off` -- all of the members it mentions in `struct flash_area` in PORTING.md.
> 
> > We could probably get rid of the fa_device_id
> 
> I wish!
> 
> But since mcuboot is getting this definition from here, we can't do that without breaking mcuboot for the reason explained above.
> 


So again that is not a job for simple PR that does switch from one API to other. There is more rework needed for both the mcuboot and the flash map API.

mbolivar-nordic commented:
>  I want to improve the current solution til we figure out what to do next and how will we approach the problem.

I think this is not a good idea for a stable API. I made a pretty detailed proposal above and it seemed like we agreed on at least one of the options; can we figure out the right thing to do instead of just merging this?

We seem to all agree that the inclusion of `struct flash_area` in the zephyr flash_map.h API is "leaking" mynewt implementation details used by a single application (mcuboot) into something that is supposed to be a generic API. I understand the goal of resolving the devices at build time, but I disagree with the proposal to do that by further mixing zephyr and mynewt by extending this structure in this way.
